api/api-doc/gossiper.json

```diff
@@ -22,6 +22,14 @@
                   "application/json"
                ],
                "parameters":[
+                  {
```

Comment:
Why push the complexity on the user of the API that may not even know the number of shards.  Just return an answer after all shards have the same value.

Reply:
> Just return an answer after all shards have the same value.

1. What if they never converge?
2. How to do this in a backward compatible way? The API should return the same result as it did unless the user provides additional (optional) parameters. Maybe a boolean flag to wait for convergence?
3. How to implement this? Querying shards periodically and checking if the results are the same?

Reply:
> > Just return an answer after all shards have the same value.
> 
>     1. What if they never converge?

If they never converge you will never know if the node is dead or alive. How it helps to be able to specify a shard if the result never converges?

> 
>     2. How to do this in a backward compatible way? The API should return the same result as it did unless the user provides additional (optional) parameters. Maybe a boolean flag to wait for convergence?

I think the previous behaviour was just wrong and the suggestion will fix it. It is not a non backwards compatible change, but a bug fix.
 
> 
>     3. How to implement this? Querying shards periodically and checking if the results are the same?

Can we do it in the api handler itself? Or are you asking how we do it there? Then the answer is "yes".

Reply:
> If they never converge you will never know if the node is dead or alive. 

The failure detector makes decisions on shard 0, then replicates it to other shards eventually. You can query shard 0 for the latest decision.

We can probably assume that it does converge eventually. A problem could happen only if the node frequently changes between up/down. This shouldn't be that frequent.

> I think the previous behaviour was just wrong and the suggestion will fix it. It is not a non backwards compatible change, but a bug fix.

Why do you think it was wrong?

Reply:
> > If they never converge you will never know if the node is dead or alive.
> 
> The failure detector makes decisions on shard 0, then replicates it to other shards eventually. You can query shard 0 for the latest decision.

Until entire node does not know the latest state the state is not relevant as seen by the problem the patch series fixes.

> 
> We can probably assume that it does converge eventually. A problem could happen only if the node frequently changes between up/down. This shouldn't be that frequent.
> 
> > I think the previous behaviour was just wrong and the suggestion will fix it. It is not a non backwards compatible change, but a bug fix.
> 
> Why do you think it was wrong?

Because we are having this patch.

Reply:
The patch only proves that there exists a use case which needs shards to converge, it doesn't prove that all use cases need it.

Since I'm not able to list all use cases, I can only come up with examples, I'd prefer to have a more flexible API.

Reply:
> The patch only proves that there exists a use case which needs shards to converge, it doesn't prove that all use cases need it.

What is the user case for knowing that one of the shards thinks that a node is up?

> 
> Since I'm not able to list all use cases, I can only come up with examples, I'd prefer to have a more flexible API.

That's the case where a common sense needs to be applied. What is the down side of waiting for convergence?

Reply:
> What is the user case for knowing that one of the shards thinks that a node is up?

Someone might be interested only in the latest information known by the node (which appears on shard 0 first), but I don't know any specific use case. I don't disagree with taking a common sense approach, but it's not zero cost.

> That's the case where a common sense needs to be applied. What is the down side of waiting for convergence?

- it might take longer, especially if there are a lot of shards; if there is a use case where it's sufficient to know the latest information only without needing it to be replicated to all shards yet, that use case might suffer from unnecessary waiting. But maybe it's insignificant.
-  it's more complex to implement than this